Converters/Civil3d/Speckle.Converters.Civil3dShared/Helpers/PropertySetBaker.cs

```diff
@@ -0,0 +1,285 @@
+using Microsoft.Extensions.Logging;
+using Speckle.Converters.Common;
+using Speckle.Sdk;
+using Speckle.Sdk.Models;
+
+namespace Speckle.Converters.Civil3dShared.Helpers;
+
+/// <summary>
+/// Helper class to bake property sets to entities on receive.
+/// </summary>
+public class PropertySetBaker
+{
+  private readonly IConverterSettingsStore<Civil3dConversionSettings> _settingsStore;
+  private readonly ILogger<PropertySetBaker> _logger;
+  private readonly PropertyHandler _propertyHandler;
+
+  public PropertySetBaker(
+    IConverterSettingsStore<Civil3dConversionSettings> settingsStore,
+    ILogger<PropertySetBaker> logger
+  )
+  {
+    _settingsStore = settingsStore;
+    _logger = logger;
+    _propertyHandler = new PropertyHandler();
+  }
+
+  /// <summary>
+  /// Try to bake property sets from a Speckle object to an AutoCAD entity.
+  /// </summary>
+  /// <param name="entity">The target entity.</param>
+  /// <param name="sourceObject">The source Speckle object containing property set data.</param>
+  /// <param name="tr">The active transaction to use for database operations.</param>
+  public bool TryBakePropertySets(ADB.Entity entity, Base sourceObject, ADB.Transaction tr)
+  {
+    if (sourceObject["properties"] is not Dictionary<string, object?> properties)
+    {
+      return false;
+    }
+
+    if (!properties.TryGetValue("Property Sets", out var propertySetsObj))
+    {
+      return false;
+    }
+
+    if (propertySetsObj is not Dictionary<string, object?> propertySets || propertySets.Count == 0)
+    {
+      return false;
+    }
+
+    try
+    {
+      foreach (var propertySet in propertySets)
+      {
+        string setName = propertySet.Key;
+        object? setDataObj = propertySet.Value;
+
+        if (setDataObj is not Dictionary<string, object?> setData)
+        {
+          _logger.LogWarning("Property set {SetName} has invalid data format", setName);
+          continue;
+        }
+
+        if (!TryBakePropertySet(entity, setName, setData, tr))
+        {
+          _logger.LogWarning("Failed to bake property set {SetName} onto entity", setName);
+        }
+      }
+
+      return true;
+    }
+    catch (Exception ex) when (!ex.IsFatal())
+    {
+      _logger.LogError(ex, "Failed to bake property sets onto entity {Handle}", entity.Handle);
+      return false;
+    }
+  }
+
+  private bool TryBakePropertySet(
+    ADB.Entity entity,
+    string setName,
+    Dictionary<string, object?> setData,
+    ADB.Transaction tr
+  )
+  {
+    try
+    {
+      ADB.ObjectId propertySetDefId = GetOrCreatePropertySetDefinition(setName, setData, tr);
+
+      if (propertySetDefId.IsNull)
+      {
+        return false;
+      }
+
+      // Check if property set already exists on the object
+      if (ObjectHasPropertySet(entity, propertySetDefId))
```

Comment:
This case should never hit - ok with throwing an exception here if it does. When do you expect a property set to be updated?